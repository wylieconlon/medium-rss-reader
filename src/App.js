import React, { Component } from 'react';

import BlogContainer from './BlogContainer';

class App extends Component {
  constructor(props) {
    super(props);

    this._parser = new DOMParser();

    this.state = {
      feedName: 'the-atlantic',
      isLoading: false,
      hasError: false,
      rssFeedDocument: null
    };
  }

  render() {
    return <div className="app">
      <header>
        <label>
          medium.com/
          <input
            type="text"
            value={this.state.feedName}
            onChange={this.onFeedNameChange.bind(this)}
          />
        </label>
        <button onClick={this.onFeedSelect.bind(this)}>
          Load Feed
        </button>
      </header>

      {this.state.hasError && <div className="error">
        Could not load feed at {this.state.feedName}
      </div>}

      {this.state.isLoading && <div className="loading">
        <div /><div /><div />
      </div>}

      {!this.state.isLoading && this.state.rssFeedDocument &&
        <BlogContainer rss={this.state.rssFeedDocument} />
      }
    </div>;
  }

  onFeedNameChange(ev) {
    this.setState({
      feedName: ev.target.value,
      hasError: false,
    });
  }

  onFeedSelect(ev) {
    this.setState({ isLoading: true });
    const getFeed = fetch(`/rss/${this.state.feedName}`, {
      method: 'POST',
    })
    .then((response) => {
      response.text().then((value) => {
        const xml = this._parser.parseFromString(value, 'application/xml');

        this.setState({ isLoading: false });

        if (xml.querySelector('parsererror')) {
          this.setState({ hasError: true, rssFeedDocument: null });
        } else {
          this.setState({ rssFeedDocument: xml });
        }
      });
    });
  }
}

export default App;
