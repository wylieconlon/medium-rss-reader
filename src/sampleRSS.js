export default `
<?xml version="1.0" encoding="UTF-8"?><rss xmlns:dc="http://purl.org/dc/elements/1.1/" xmlns:content="http://purl.org/rss/1.0/modules/content/" xmlns:atom="http://www.w3.org/2005/Atom" version="2.0" xmlns:cc="http://cyber.law.harvard.edu/rss/creativeCommonsRssModule.html">
    <channel>
        <title><![CDATA[Blog Name - Medium]]></title>
        <description><![CDATA[Blog Description - Medium]]></description>
        <link>https://medium.com/blog-url</link>
        <image>
            <url>https://cdn-images-1.medium.com/proxy/1*TGH72Nnw24QL3iV9IOm4VA.png</url>
            <title>Compass True North - Medium</title>
            <link>https://medium.com/compass-true-north?source=rss----2a2f97cb73e3---4</link>
        </image>
        <generator>Medium</generator>
        <lastBuildDate>Wed, 17 Oct 2018 15:43:41 GMT</lastBuildDate>
        <atom:link href="https://medium.com/feed/compass-true-north" rel="self" type="application/rss+xml"/>
        <webMaster><![CDATA[yourfriends@medium.com]]></webMaster>
        <atom:link href="http://medium.superfeedr.com" rel="hub"/>
        <item>
            <title><![CDATA[Working with AWS & Agent Feedback to build a better product]]></title>
            <link>https://medium.com/compass-true-north/working-with-aws-agent-feedback-to-build-a-better-product-c3009b6e6184?source=rss----2a2f97cb73e3---4</link>
            <guid isPermaLink="false">https://medium.com/p/c3009b6e6184</guid>
            <category><![CDATA[product]]></category>
            <category><![CDATA[expansion]]></category>
            <category><![CDATA[compass]]></category>
            <category><![CDATA[engineering]]></category>
            <dc:creator><![CDATA[Brian Renzenbrink]]></dc:creator>
            <pubDate>Mon, 08 Oct 2018 14:55:35 GMT</pubDate>
            <atom:updated>2018-10-08T14:55:35.494Z</atom:updated>
            <content:encoded><![CDATA[<p>A few weeks back, we got to work with the folks at <a href="https://uncubed.com/">UnCubed</a> in partnership with AWS to put together a short video describing one of the early challenges in our expansion and how our infrastructure in AWS helped us work through it. Check itÂ out!</p>]]></content:encoded>
        </item>
    </channel>
</rss>
`;
