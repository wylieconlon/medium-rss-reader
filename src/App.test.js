import React from 'react';
import { shallow } from 'enzyme';

window.fetch = makeFetchMock();
import App from './App';
import sampleRSS from './sampleRSS';

function makeFetchMock() {
  return jest.fn().mockResolvedValue({
    text: jest.fn().mockResolvedValue(sampleRSS),
  });
}

it('renders without crashing', () => {
  expect(shallow(<App />).html()).toMatchSnapshot();
});

it('loads an rss document', async () => {
  const app = shallow(<App />);

  app.find('button').simulate('click');

  await window.fetch();

  expect(app.state('isLoading')).toEqual(false);
  expect(app.state('rssFeedDocument')).toBeTruthy();
});
