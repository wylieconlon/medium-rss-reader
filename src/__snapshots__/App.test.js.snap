// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `"<div class=\\"app\\"><header><label>medium.com/<input type=\\"text\\" value=\\"the-atlantic\\"/></label><button>Load Feed</button></header></div>"`;
